DM-15189: core butler support for raw data ingest

TallJimbo opened the pull request:
@timj, this is by no means done yet, but it occurred to me that we might be doing conflicting things with `Datastores`, so I wanted to get this out there in case you'd like me to spin it off to a new issue and merge sooner.  I also need to write some tests for it, but since you've just refactored `Datastore` tests I definitely didn't want to do that before rebasing on that work.

TallJimbo committed:
Add method to retrieve a DataUnit object from SqlRegistry.

TallJimbo commented:
Most of the code originally here was moved to DM-15212, so some earlier comments may no longer make sense.

TallJimbo committed:
Refactor Visit/Exposure entry from VisitInfo routines.

These are useful for raw data ingest, too, so move them out of
the convert package.  New home still probably isn't their
permanent one.

TallJimbo committed:
Make TopologicalSet actually behave like a Set.

This change didn't turn out to be necessary for this ticket,
but it seemed worth preserving.

TallJimbo committed:
Concrete StorageClass adjustments.

The duplication of all Exposure components in order to refine
the types of some is unfortunate, but should go away after
DM-15238.

TallJimbo committed:
Add formatter base class for raw data.

In retrospect, this needs to be modified to produce an
Exposure, not a DecoratedImage, and we can safely
remove the write method.

TallJimbo committed:
Refactor FitsExposureFormatter for readability, extensibility.

TallJimbo committed:
Add TransmissionCurve component to Exposure StorageClass

TallJimbo committed:
Make Dataset-not-found error friendlier.

TallJimbo committed:
Make certain ingested paths are relative to root.

TallJimbo committed:
Add run attribute to DatasetRef.

TallJimbo committed:
Add parameter support to Butler.get.

TallJimbo committed:
Make it possible to disable method-level transactionality.

This can be useful when you don't want an exception to start a
rollback because you're about to catch it.

TallJimbo committed:
Add run, collection keys to file templates.

TallJimbo commented:
The full ticket is finally ready for review.  In this package (as well as obs_subaru), I recommend looking at each commit separately; this package in particular just contains a lot of miscellaneous improvements necessary for making higher-level code in obs_base and obs_subaru work.

timj commented on config/storageClasses.yaml:
```diff
@@ -19,12 +19,21 @@ storageClasses:
+  Image: &Image
+    pytype: lsst.afw.image.Image
   ImageF:
+    <<: *Image
```
Since `Image` only has one field and that field is being over-ridden, I'm not sure this line does anything (and the other similar changes). Are you worried that `Image` might get extra fields?

TallJimbo replied:
Yeah, it doesn't actually do anything; I just felt that using `<<:` to indicate inheritance even when it was a no-op was a useful bit of code-as-documentation.

timj replied:
It's not too late to add a `parent:` keyword and get proper pythonic inheritance for StorageClasses (I think I can sort out print through of those class properties).

TallJimbo replied:
Let's see what happens with [DM-15238](https://jira.lsstcorp.org/browse/DM-15238) first.

timj commented on config/datastores/posixDatastore.yaml:
```diff
@@ -5,7 +5,7 @@ datastore:
   create: true
   templates:
-    default: "{datasetType}/{tract:?}/{patch:?}/{filter:?}/{camera:?}_{visit:?}"
+    default: "{collection}/{datasetType}/{tract:?}/{patch:?}/{filter:?}/{camera:?}_{visit:?}"
```
General comment on defaulting filenames: I'm a bit worried that with this in the base definition it's not possible for people to remove defaulting in their own datastores. Maybe I should allow `default: None` to disable defaulting.

timj commented on config/storageClasses.yaml:
```diff
@@ -41,31 +50,50 @@ storageClasses:
   ExposureComposite: &ExposureComposite
     <<: *Exposure
     assembler: lsst.daf.butler.assemblers.exposureAssembler.ExposureAssembler
   ExposureCompositeF:
```
This change won't be needed once #76 is merged because `ExposureCompositeF` becomes a simple label that allows you to control concrete/virtual at the butler level.

timj commented on config/storageClasses.yaml:
```diff
@@ -41,31 +50,50 @@ storageClasses:
+      apCorrMap: TablePersistableApCorr
+      coaddInputs: TablePersistableCoaddInputs
+      transmissionCurve: TablePersistableTransmissionCurve
+      metadata: PropertyList
```
`getMetadata` is documented to return a `PropertySet` though?

TallJimbo replied:
Yeah, we discussed this on the other ticket.  I'll convert to `PropertySet` here.

timj replied:
I actually left it alone in #76 since it worked regardless.

TallJimbo replied:
Ok, I think I'll stick with `PropertyList` here, too, then.

timj commented on python/lsst/daf/butler/butler.py:
```diff
@@ -266,6 +266,9 @@ def getDirect(self, ref):
             Reference to an already stored dataset.
+        parameters : `dict`
+            Additional StorageClass-defined options to control reading,
+            typically used to efficiently only a subset of the dataset.
```
Missing word between "efficiently" and "only". Maybe I should work on parameters next (inMemoryDatastore will mess up with any tests where you use parameters).

timj commented on python/lsst/daf/butler/butler.py:
```diff
@@ -274,20 +277,25 @@ def getDirect(self, ref):
+                if parameters is None:
+                    compParams = None
+                else:
+                    compParams = {k: v for k, v in parameters.items()
```
Parameters are nested by component name?

TallJimbo replied:
No, I'm just filtering out parameters that aren't supported by that component.  So, for instance, if you pass a box for a subimage when getting a virtual composite Exposure, it gets passed down to `image`, `mask`, and `variance`, but not `wcs`.

I'll add a code comment.

timj commented on python/lsst/daf/butler/butler.py:
```diff
@@ -300,6 +308,9 @@ def get(self, datasetRefOrType, dataId=None):
             argument.
+        parameters : `dict`
+            Additional StorageClass-defined options to control reading,
+            typically used to efficiently only a subset of the dataset.
```
Same missing word here.

timj commented on python/lsst/daf/butler/butler.py:
```diff
@@ -318,9 +329,12 @@ def get(self, datasetRefOrType, dataId=None):
         # Always lookup the DatasetRef, even if one is given, to ensure it is
         # present in the current collection.
         ref = self.registry.find(self.collection, datasetType, dataId)
+        if ref is None:
```
Our tests never tickled this?

TallJimbo replied:
I'm not sure; the old behavior definitely raised an exception, but it was a pretty incomprehensible one.  My goal here was just it more comprehensible.

timj commented on tests/test_templates.py:
```diff
@@ -37,7 +37,7 @@ def makeDatasetRef(self, datasetTypeName, dataId=None):
                                                              StorageClass(None))
         datasetType = self.datasetTypes[datasetTypeName]
-        return DatasetRef(datasetType, dataId)
+        return DatasetRef(datasetType, dataId, run=Run(id=2, collection="run2"))
```
I don't see a test in here for `collection` only, both `collection` and `run`, or for where `run` and `collection` are not specified in the template.

timj commented on python/lsst/daf/butler/core/fileTemplates.py:
```diff
@@ -162,6 +172,11 @@ def format(self, ref):
+            if field_name in ("run", "collection"):
+                usedRunOrCollection = True
+                if optional:
+                    raise KeyError("'run' and 'collection' may not be optional.")
```
Are you allowed to specify an optional `run` if `collection` is mandatory? (not that it's possible for `run` to be optional of course). Maybe the error here should be "can not be optional" ("may" sounds like you might be able to be optional). Do you have a test for optional `collection` below?

TallJimbo replied:
:+1: on "can be optional".  Will add a test.

TallJimbo replied:
Oh, actually, yeah, I suppose you could make one optional if the other is mandatory.  Probably best to just remove this particular check, as there is another one to verify that at least one of those is present.

timj commented on python/lsst/daf/butler/formatters/fitsExposureFormatter.py:
```diff
@@ -33,6 +33,53 @@ class FitsExposureFormatter(Formatter):
+        tiny = self.readFull(fileDescriptor, parameters)
+        return fileDescriptor.storageClass.assembler().getComponent(tiny, component)
+
+    def readFull(self, fileDescriptor, parameters=None):
```
You are defining the parameters for `get` to be any parameters understood by the Exposure constructor?

TallJimbo replied:
Not at the moment - just `bbox` and `origin`.  I suppose we need a place to document this at the StorageClass level.  I'm open to ideas.

timj commented on python/lsst/daf/butler/formatters/fitsRawFormatterBase.py:
```diff
@@ -0,0 +1,139 @@
+__all__ = ("FitsRawFormatterBase",)
+
+
+class FitsRawFormatterBase(FitsExposureFormatter, metaclass=ABCMeta):
```
All of the methods here need parameters and returns sections to the numpydocs.

timj commented on python/lsst/daf/butler/formatters/fitsExposureFormatter.py:
```diff
@@ -33,6 +33,53 @@ class FitsExposureFormatter(Formatter):
 
     parameters = frozenset(("bbox", "origin"))
 
+    def readImageComponent(self, fileDescriptor, component):
```
Please add parameters/returns sections to docstrings.

timj commented on python/lsst/daf/butler/instrument.py:
```diff
@@ -65,6 +71,76 @@ def _addPhysicalFilters(self, registry):
+        A dictionary containing all fields in the Exposure table.
+    """
+    avg = visitInfo.getDate()
+    begin = DateTime(int(avg.nsecs(DateTime.TAI) - 0.5E9*visitInfo.getExposureTime()), DateTime.TAI)
```
You mean we can't get the value of `DATE-OBS` from the object but need to guess it by subtracting half the exposure time?

TallJimbo replied:
AFAIK, yes.  I don't know much about `VisitInfo`.  I stole this recipe from `obs.base.makeRawVisitInfo.offsetDate`, and I'd like to replace both that and `VisitInfo` itself, but this seemed like an easy way to get the correct (or at least consistent with Gen2) answer.

timj reviewed (approved):
In general looks good. I have quite a few comments.

TallJimbo commented:
I think I've addressed all comments (and rebased on latest master).

One question from that rebase (which I should have asked on a previous review, probably): do we still need the `ExposureComposite` StorageClasses?

timj commented:
Re ExposureComposite: I think we should move it to the test configs so that the tests can try a virtual and concrete composite.

TallJimbo commented:
> I think we should move it to the test configs so that the tests can try a virtual and concrete composite.

I gave this a quick try on this ticket, but it's not easy because aliases apparently can't be used outside the file where they're defined (so you can't easily make `ExposureComposite` inherit attributes from `Exposure` when they're in different files).

TallJimbo committed:
Make addDataset optionally recursive.

timj commented:
I was worrying about that cross-yaml file inheritance this morning. Would trivially be fixed if we had true python inheritance as well.  Leave ExposureComposite in there for now -- I really do need to make a test for Exposure that uses virtual and one that uses concrete.
